Mobile rl

dHonerkamp opened the pull request:
Hi everyone,

This is an initial version with our approach on mobile manipulation based on our paper (https://arxiv.org/abs/2101.05325).
It's not completely ready to be merged yet, but should already include all the main parts.

- It implements the LearnerRL interface
- Formatted according to PEP-8, .clang-format
- Most unnecessary functionality should already be removed
- Includes a first version of the documentation, including examples to train and to evaluate provided checkpoints

But there are also a few questions from my side. Mainy because this is a project that relies on a python3-based interface for the user and an environment implemented in c++, which additionally draws on functionality from ROS (mainly moveit).

- Atm I am keeping the c++ src and header files within the module, combined with it's own CMakeLists.txt. (i.e. within `src/control/mobile_manipulation/`). Is that appropriate?
- How should I define ROS and c++ dependencies? The module provides environments for several robotic platforms (PR2, Tiago, HSR). This means to compile or run the module the user needs (i) a ROS installation (developped and tested for melodic) (ii) a separate catkin_ws for each robot (iii) launch a launchfile before running the python scripts. Due to this I feel like it makes sense to not require every user of other openDR modules to install these, but rather have them specified specific to this module. Some of robot specific dependencies should furthermore be compiled in separate catkin_workspaces. The model checkpoints are tiny (3x 3MB) and currently directly located in the git repo. Is that ok for such small files?
- Licenses: this module includes slightly modified launchfiles from openly available ROS packages (`~/robots_worlds/[pr2/hsr/tiago]`). Do these have to be marked or treated specially somehow?
- This was developped as part of WP5.2 Deep Navigation. As there is no navigation folder and this approach can be seen as a combination of navigation and control I have located it within control for now. Let me know in case I should move it elsewhere

Any help on the above would be much appreciated. Other comments on what is already in here are also welcome as well.

Some remaining todo's for myself to remember:
- update checkpoints
- test that gazebo evaluation works
- test that examples in readme work

passalis commented on src/control/mobile_manipulation/include/mobile_manipulation_rl/worlds.h:
```diff
@@ -0,0 +1,60 @@
+#pragma once
```
I am not 100% sure about the C/C++ coding style (perhaps @stefaniapedrazzi can answer this), but I think that we should have a uniform way of structuring header files. Some files uses traditional include guards, while this one uses a *pragma* directive.

dHonerkamp replied:
Happy to adjust if someone can confirm which convention to use.

passalis replied:
I do not have a preference. Theoretically, the traditional include guards are supported by every compiler, so they are more portable (so I would tend towards this for this reason). However, I also think that every major compiler also supports pragma directives so this is probably not a real concern.

dHonerkamp replied:
Done, using traditional guards consistently now

passalis commented on src/control/mobile_manipulation/mobileRL/envs/robotenv.py:
```diff
@@ -0,0 +1,166 @@
+            self._env = RobotPR2(*args)
+        elif env == 'tiago':
+            self._env = RobotTiago(*args)
+        # elif env == 'hsr':
```
Please removed unused parts of code, unless there is a reason for keeping them.

passalis commented on src/control/mobile_manipulation/mobile_manipulation_learner.py:
```diff
@@ -0,0 +1,247 @@
+            self.load(restore_model_path)
+
+    @staticmethod
+    def _download_pretrained(save_path: str, robot_name: str):
```
Even though the signature of this function has not been specified in the first version of specifications, I think it would be nice if we could stick to an existing way of implementing this (e.g., https://github.com/tasostefas/opendr_internal/blob/master/src/opendr/perception/pose_estimation/lightweight_open_pose/lightweight_open_pose_learner.py#L906 ) in order to avoid having to update this later. So, it would be very nice if you could rename and add any arguments (if needed) to follow the function signature provided by the rest of the implementations.

dHonerkamp replied:
This is a private helper function as indicated by the underscore. The load() function adheres to the BaseLearner Api and is the one that should be called by any user. I am not sure how making this public would provide additional functionality over the load function and in the worst case seems to me like it would confuse users, as there is also no public download functionality in any of the base classes?

passalis replied:
What we have found out during the implementation is that even though there is no public *download()* function specified in D2.1, most method needs one to download a pretrained model. If this is the case and you download a pre-trained model, then it would be beneficial to provide a public *download()* method to follow the interface of the rest of the methods (even through it is not included in the base learner). This is indeed not yet specified in D2.1, but it will most probably be in the next version, since virtually all methods provide this functionality.

dHonerkamp replied:
I see. I have adjusted the signature.

passalis commented on src/control/mobile_manipulation/mobile_manipulation_learner.py:
```diff
@@ -0,0 +1,247 @@
+                                                                                    global_step=self.stable_bl_agent.num_timesteps,
+                                                                                    verbose=2)
+        env.clear()
+        return episode_rewards, episode_lengths, metrics, name_prefix
```
Please return a dictionary with these values, as described in the specifications.

dHonerkamp replied:
Done

passalis commented on src/control/mobile_manipulation/package.xml:
```diff
@@ -0,0 +1,71 @@
+  <version>0.0.0</version>
+  <description>The Kinematic Feasibility RL package</description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
```
Please remove the comments.

dHonerkamp replied:
Done

passalis commented on src/control/mobile_manipulation/robots_world/hsr/hsr_analytical.launch:
```diff
@@ -0,0 +1,61 @@
+
+  <!-- Start the actual move_group node/action server -->
+  <node name="move_group" launch-prefix="" pkg="moveit_ros_move_group" type="move_group" respawn="false" output="screen" args="">
+    <!-- Set the display variable, in case OpenGL code is used internally -->
```
Please remove the template comments (keep only those that are not the standard ones).

dHonerkamp replied:
Removed those that do not add information

passalis commented on projects/control/mobile_manipulation/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+## Compile as C++11, supported in ROS Kinetic and newer
+# add_compile_options(-std=c++11)
+
+## Find catkin macros and libraries
```
Please remove template comments.

dHonerkamp replied:
Done

passalis commented on projects/control/mobile_manipulation/mobile_manipulation_demo.py:
```diff
@@ -0,0 +1,259 @@
+# Copyright 1996-2020 OpenDR European Project
```
Why do we need two licenses?

passalis commented on src/control/mobile_manipulation/include/mobile_manipulation_rl/worlds.h:
```diff
@@ -0,0 +1,60 @@
+};
+
+class GazeboWorld : public BaseWorld {
+private:
```
Please removed unused code.

dHonerkamp replied:
Done

passalis commented on src/control/mobile_manipulation/robots_world/tiago/tiago_analytical.launch:
```diff
@@ -0,0 +1,90 @@
+
+  <arg name="capabilities" default=""/>
+  <arg name="disable_capabilities" default=""/>
+  <!-- load these non-default MoveGroup capabilities (space seperated) -->
```
Please removed unused definitions, unless needed.

dHonerkamp replied:
Removed

passalis commented on src/control/mobile_manipulation/src/gripper_planner/gaussian_mixture_model.cpp:
```diff
@@ -0,0 +1,660 @@
+      tf_Mu_base_i.getOrigin().x(), tf_Mu_base_i.getOrigin().y(), tf_Mu_base_i.getOrigin().z(), tf_Mu_base_i.getRotation().x(),
+      tf_Mu_base_i.getRotation().y(), tf_Mu_base_i.getRotation().z(), tf_Mu_base_i.getRotation().w();
+
+    // restrict torso height to bounds
```
Please removed unused code, unless needed.

dHonerkamp replied:
removed

passalis commented on src/control/mobile_manipulation/robots_world/tiago/tiago.urdf.xacro:
```diff
@@ -0,0 +1,147 @@
+<?xml version="1.0"?>
```
I am not sure that this file can be included in the toolkit. Perhaps PAL robotics can be contacted if we need to change the license.

dHonerkamp replied:
I am happy to contact PAL. What would the desired resolution be in this case, i.e. is the question whether we would be allowed to change the licence on this file to Apache 2.0 or simply whether PAL gives us the permission to include this file with this license?

passalis replied:
I think we need this file with Apache 2.0 license in order to be included in the toolkit. So  most probably we need to either change the license or a version of the file with a different license.  I think this shouldn't be an issue, since PAL is a partner in this project.

dHonerkamp replied:
I have contacted PAL robotics about this

passalis commented on src/control/mobile_manipulation/src/pybindings.cpp:
```diff
@@ -0,0 +1,125 @@
+    .def("world_to_relative_tf", &RobotTiago::worldToRelativeTf, "worldToRelativeTf.")
+    .def("publish_marker", &RobotTiago::publishMarker, "publishMarker.");
+
+  //    py::class_<RobotHSR>(m, "RobotHSR")
```
Please remove unused code portions.

dHonerkamp replied:
See comment on HSR above

passalis reviewed (commented):
Thanks for the contribution! 

I have made some comments directly on the files. My main concern is whether this PR will pass the tests on the integration server, since it runs Ubuntu 20.04, while this PR seems to depend on ROS melodic. Also, there is a file (https://github.com/tasostefas/opendr_internal/blob/mobileRL/src/control/mobile_manipulation/robots_world/tiago/tiago.urdf.xacro) that seems to have a license incompatible with Apache ,

dHonerkamp commented on src/control/mobile_manipulation/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+cmake_minimum_required(VERSION 2.8.12)
+project(mobile_manipulation_rl)
+
+## Compile as C++11, supported in ROS Kinetic and newer
```
Done

dHonerkamp commented on projects/control/mobile_manipulation/package.xml:
```diff
@@ -0,0 +1,29 @@
+  <version>0.0.0</version>
+  <description>The Kinematic Feasibility RL Demo package</description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
```
Done

dHonerkamp commented:
How do licences in cpp files have to look like?

First the tests were complaining that I am not using this licence:
```
* Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```
After formatting it exactly like the above, it is now complaining that they should like below. Which was the firs formatting I used before it told me to use the above formatting.

```
======================================================================
FAIL: test_sources_have_license (test_license.TestLicense)
Test that sources have the license.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/opendr_internal/opendr_internal/tests/test_license.py", line 140, in test_sources_have_license
    self.assertTrue(
AssertionError: False is not true : Source file "/home/runner/work/opendr_internal/opendr_internal/src/control/mobile_manipulation/src/robot_tiago.cpp" doesn't contain the correct Apache 2.0 License:
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

Please let me know how to pass these tests.

stefaniapedrazzi commented:
> How do licences in cpp files have to look like?
> 
> First the tests were complaining that I am not using this licence:
> 
> ```
> * Licensed under the Apache License, Version 2.0 (the "License");
>  * you may not use this file except in compliance with the License.
>  * You may obtain a copy of the License at
>  *
>  *     http://www.apache.org/licenses/LICENSE-2.0
>  *
>  * Unless required by applicable law or agreed to in writing, software
>  * distributed under the License is distributed on an "AS IS" BASIS,
>  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>  * See the License for the specific language governing permissions and
>  * limitations under the License.
>  */
> ```
> 
> After formatting it exactly like the above, it is now complaining that they should like below. Which was the firs formatting I used before it told me to use the above formatting.
> 
> ```
> ======================================================================
> FAIL: test_sources_have_license (test_license.TestLicense)
> Test that sources have the license.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/opendr_internal/opendr_internal/tests/test_license.py", line 140, in test_sources_have_license
>     self.assertTrue(
> AssertionError: False is not true : Source file "/home/runner/work/opendr_internal/opendr_internal/src/control/mobile_manipulation/src/robot_tiago.cpp" doesn't contain the correct Apache 2.0 License:
> // Licensed under the Apache License, Version 2.0 (the "License");
> // you may not use this file except in compliance with the License.
> // You may obtain a copy of the License at
> //
> //     http://www.apache.org/licenses/LICENSE-2.0
> //
> // Unless required by applicable law or agreed to in writing, software
> // distributed under the License is distributed on an "AS IS" BASIS,
> // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> // See the License for the specific language governing permissions and
> // limitations under the License.
> ```
> 
> Please let me know how to pass these tests.

They should have this format:
https://github.com/tasostefas/opendr_internal/blob/55b60c4ec86df5db844d1706161ccbb05c9dece3/tests/test_license.py#L38-L48

dHonerkamp commented on docs/reference/mobile-manipulation.md:
`````diff
@@ -0,0 +1,269 @@
+roslaunch mobile_manipulation_rl [pr2,tiago]_analytical
+````
+
+The environment was tested for Ubuntu 18.04 and ROS melodic and assumes this version in its dependencies installation.
`````
I am currently not able to run the unit tests, see comment below

passalis replied:
> @passalis thank you for you comments. I believe I have addressed all of them, could you maybe take a look at my replies?

Thanks for the changes! They seem fine, I have responded to the ones that require a further action and are not blocked.

dHonerkamp commented:
Thanks @stefaniapedrazzi  . I believe I am now using the format you linked, but I now again receive the initial error for the header files:
```
======================================================================
FAIL: test_sources_have_license (test_license.TestLicense)
Test that sources have the license.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/opendr_internal/opendr_internal/tests/test_license.py", line 134, in test_sources_have_license
    self.assertTrue(
AssertionError: False is not true : Source file "/home/runner/work/opendr_internal/opendr_internal/src/opendr/control/mobile_manipulation/include/mobile_manipulation_rl/robot_tiago.h" doesn't contain the correct Apache 2.0 License:
* Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */`
```
Do I have to use a different format for .h and .cpp files?

stefaniapedrazzi commented:
> Do I have to use a different format for .h and .cpp files?

CPP code should use `.hpp` extension for header files and `.cpp` extension for source files.
This way it is clear that it is CPP code and not C.

However, in `.hpp`/`.cpp` the license is exptected using the comment `\\`.
In `.h`/`.c`the license is expected using the multine comment style `/* ... */`

dHonerkamp commented:
Thank you @stefaniapedrazzi . Unfortunately I am now stuck at the tests failing on typing, as described in #42 . Is there any resolution to this?

stefaniapedrazzi commented:
> Thank you @stefaniapedrazzi . Unfortunately I am now stuck at the tests failing on typing, as described in #42 . Is there any resolution to this?

No. For the moment this is not supported.

dHonerkamp commented:
Thanks for the reply. We are using typing throughout every single python function, so I guess this will block integration for the moment

dHonerkamp commented:
@passalis thank you for you comments. I believe I have addressed all of them, could you maybe take a look at my replies? 

The only one not yet addressed is the ROS version, unfortunately as you can see in the discussion with Stefania above, it is currently not possible to pass the unit tests due to python typings.

passalis commented:
> @passalis thank you for you comments. I believe I have addressed all of them, could you maybe take a look at my replies?

Thanks for the changes! They seem fine, I have responded to the ones that require a further action and are not blocked.

dHonerkamp commented:
HEY @jelledouwe @stefaniapedrazzi  @passalis what is going on with this forced pushed into my PR? Please explain, this is highly risky behavior and could cost me losing progress!

dHonerkamp commented:
@jelledouwe In my understanding it is rule number 1 to (i) not make any forced pushes into collaborative repositories (ii) not push onto PRs of other people.

I was just working on this.

Can someone please also repoen this PR, I don't seem to have the rights to do so.

dHonerkamp commented:
@jelledouwe I will reinstate this with my local version, which I luckily still have, in 5 min unless you can give me a very good reason. Either way I would like an explanation.

jelledouwe commented:
Dear @dHonerkamp,

I am terribly sorry about what happened. You are completely right and I should never have done this. It was never my intention to do this. I would really like to tell explain it to you in person via Teams. Please let me know if you are interested in this. Let me know if I can do anything to fix it. It will never happen again. I hope that you didn't lose any work. Again, my sincere apologies.

dHonerkamp commented:
Hi @jelledouwe,
I appreciate the apology. Luckily I still had all changes locally. As such let's just book this under honest mistakes. Just as a general recommendation I would generally abstain from any forced pushes, even onto your own branches.

@passalis Can you repopen the PR? I addressed your comments. There should only be two things left:
- the license on the tiago urdf -> contacted PAL
- the unittests -> currently blocked by the missing linting support for typing

jelledouwe commented:
Dear @dHonerkamp,

Thank you for your kind response. I am relieved to hear that you still have all changes. I will definitely stay away from forced pushes in the future.

passalis commented:
Hi @dHonerkamp,

We cannot re-open the PR due to the force push. My suggestion would be to try following the instructions in https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c .

So, I think we need to restore f75e8da (the hash of the latest push before this PR was closed) and after this we can reopen the issue and push again any changes made after this point (if any). Can you check if this works for you?

@stefaniapedrazzi Is this the preferred way to do this or is there any other solution?

stefaniapedrazzi commented:
The best procedure is to create a new branch based on the master and commit again the files modified in this PR.
This way we are sure that nothing is messed up and that the new branch is clean and doesn't contain any unrelated changes.

I have a local copy of this branch at f75e8da8096f0a1047c04a5e456d0d152979fc0b.
If you need I can create (probably tomorrow) a new branch corresponding to that state.

dHonerkamp commented:
This sounds to me like it is safer to just open a new PR from the restored branch. I have done so #115 and linked to the discussion here. I hope that this approach is ok with you.
